src/libsyntax_pos/lib.rs

```diff
@@ -60,22 +63,31 @@ pub type FileName = String;
 /// that the length of the span = hi - lo; there may be space in the BytePos
 /// range between files.
 #[derive(Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
-pub struct Span {
-    #[unstable(feature = "rustc_private", issue = "27812")]
-    #[rustc_deprecated(since = "1.21", reason = "use getters/setters instead")]
-    pub lo: BytePos,
-    #[unstable(feature = "rustc_private", issue = "27812")]
-    #[rustc_deprecated(since = "1.21", reason = "use getters/setters instead")]
-    pub hi: BytePos,
+struct SpanData {
+    lo: BytePos,
+    hi: BytePos,
     /// Information about where the macro came from, if this piece of
     /// code was created by a macro expansion.
-    #[unstable(feature = "rustc_private", issue = "27812")]
-    #[rustc_deprecated(since = "1.21", reason = "use getters/setters instead")]
-    pub ctxt: SyntaxContext,
+    ctxt: SyntaxContext,
 }
 
-#[allow(deprecated)]
-pub const DUMMY_SP: Span = Span { lo: BytePos(0), hi: BytePos(0), ctxt: NO_EXPANSION };
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
```

Comment:
@michaelwoerister 
I'm not sure if `Span` can derive `Hash` or not (for incremental, etc).
Is hashing the 32-bit "index" enough for interned spans, or actual data must be hashed?

Reply:
Incremental uses `HashStable` anyway.

Reply:
i.e. this impl:

https://github.com/rust-lang/rust/blob/ef227f5ffe438783db0add58208d79dd84158c06/src/librustc/ich/hcx.rs#L234

Reply:
Yes, as @arielb1 says. We keep hashing for incr. comp. separate because often it's a lot expensive than what we need for a regular hash table and it often requires additional contextual information. So you don't have to worry about it.

You might want to provide a specialized implementation that hashes just one 32 bit value instead for bytes though (unless we do that anyway). But probably not worth the trouble.